controlpanel/kubeapi/views.py

```diff
@@ -0,0 +1,53 @@
+import logging
+import os
+
+from django.conf import settings
+from djproxy.views import HttpProxy
+from kubernetes import client, config
+
+# This patch fixes incorrect base64 padding in the Kubernetes Python client.
+# Hopefully it will be fixed in the next release.
+from controlpanel.kubeapi import oidc_patch
+
+
+log = logging.getLogger(__name__)
+
+
+if 'KUBERNETES_SERVICE_HOST' in os.environ:
+    config.load_incluster_config()
+
+else:
+    config.load_kube_config()
+
+
+
+class KubeAPIAuthMiddleware(object):
+    """
+    Add user's token to the Authorization header
+    """
+
+    def process_request(self, proxy, request, **kwargs):
+        if settings.ENABLED["k8s_rbac"]:
+            _, token = request.META["HTTP_AUTHORIZATION"].split(" ", 1)
+            auth = f"Bearer {token}"
+
+        else:
+            auth = client.Configuration().api_key["authorization"]
+
+        log.debug("auth = %s", auth)
+
+        kwargs["headers"]["Authorization"] = auth
+        return kwargs
+
+
+class KubeAPIProxy(HttpProxy):
+    """
+    Proxy requests to the Kubernetes cluster API
+    """
+
+    base_url = client.Configuration().host
+
+    # Without this, we get SSL: CERTIFICATE_VERIFY_FAILED
+    verify_ssl = client.Configuration().ssl_ca_cert
+
+    proxy_middleware = ["controlpanel.kubeapi.views.KubeAPIAuthMiddleware"]
```

Comment:
### 2. Move Kubernetes API proxy into its own app
> Lines 43 - 53

I junked the existing proxy view function we had in favour of [djproxy](https://github.com/thomasw/djproxy) on the principle that it is always better to use off-the-shelf code than maintain and worry about the correctness of our own implementation.

As a result, the Kubernetes API proxy is reduced to a few lines of code configuring the proxy View class, and a middleware object that inserts the request user's authentication token into proxied requests to the Kubernetes API.

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-8c49a844762f3e9fdeb511f2a34ca23eR77)

Reply:
*Disclaimer*: I'm still wrapping my head around the whole thing.

Specifically, [I don't see the bit that we were doing where we were stripping the first few characters of the path](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-85ef3a34be3164c0b669e382d500904dL23) of the requests sent to k8s (because they were maid to `/k8s/`). 

But **the most important question is**: do we still need the proxy? I haven't look the whole thing but I see some places where k8s requests are made through the Python client. I can't really remember, sorry if this is really a stupid question.

Reply:
I've [changed the URLs for the k8s proxy](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-0ad9842f86b8342a9266c007338496adR9). The base URL path for the `kubeapi` proxy app is set in `controlpanel/urls.py`:
```python
path("api/k8s/", include("controlpanel.kubeapi.urls"))
```
and the `kubeapi` URL config looks like
```python
path("<path:url>", views.KubeAPIProxy.as_view(), name="proxy")
```
which means anything after `/api/k8s/` in the URL is the proxied URL.

Reply:
Regarding whether we still need the K8s proxy - I'm not sure. I didn't want to remove it in this PR, in case it is being used - _but_ I changed the URLs, so that would probably break anything that is expecting `/k8s/` instead of `/api/k8s/`.

Reply:
Right, so if I request `/api/k8s/foo` the request to k8s will be `/foo` 👍

Reply:
Regarding the use of the proxy to k8s. I think we may have to use it for all requests to k8s from the UI.
The UI now makes requests using the Python client but that uses the CP service account.

Is this what we want? Before this PR the user was using the same JWT token (by design) to make requests to both API and k8s API which meant:
1. we know who performs what
2. if there is a bug in the API/UI and a user is able to perform an operation which they're not supposed to do the RBAC in the k8s API would stop it (user would get a `403`)

If UI uses Python client and CP service account these two things will not be true anymore.
The CP would perform things on behalf of the users but the CP has more permissions than normal users and this can potentially lead to problems.

CC: @r4vi

Reply:
Ah yes, I didn't mean to change that functionality. I will fix it.